Support variadic returns in Schema's operator<<

houseroad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23422 [WIP] Add CI job to check function schemas' backward compatiblity
* #23409 Add isBackwardCompatibleWith for Argument and FunctionSchema
* #23208 expose parse_schema and __eq__ function to python and add round trip tests
* #23207 Expose an API to iterate all the registered operators
* #23206 Align AliasInfo's operator<< with FunctionSchema
* #23205 Support IntList in Dict's shalloEquals
* **#23204 Support variadic returns in Schema's operator<<**

old: prim::PythonOp(...) ->
new: prim::PythonOp(...) -> ...

Differential Revision: [D16433592](https://our.internmc.facebook.com/intern/diff/D16433592/)

houseroad committed:
Support variadic returns in Schema's operator<<

old: prim::PythonOp(...) ->
new: prim::PythonOp(...) -> ...

Differential Revision: [D16433592](https://our.internmc.facebook.com/intern/diff/D16433592/)

houseroad committed:
Update on "Support variadic returns in Schema's operator<<"


old: prim::PythonOp(...) ->
new: prim::PythonOp(...) -> ...

Differential Revision: [D16433592](https://our.internmc.facebook.com/intern/diff/D16433592/)

houseroad committed:
Update on "Support variadic returns in Schema's operator<<"


old: prim::PythonOp(...) ->
new: prim::PythonOp(...) -> ...

Differential Revision: [D16433592](https://our.internmc.facebook.com/intern/diff/D16433592/)

houseroad committed:
Update on "Support variadic returns in Schema's operator<<"


old: prim::PythonOp(...) ->
new: prim::PythonOp(...) -> ...

Differential Revision: [D16433592](https://our.internmc.facebook.com/intern/diff/D16433592/)

dzhulgakov commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -29,13 +29,17 @@ inline std::ostream& operator<<(std::ostream& out, const FunctionSchema& schema)
-  } else if (schema.returns().size() > 1) {
+
+  const auto& returns = schema.returns();
+  if (schema.is_varret()) {
```
if varret is analogous to vararg, it should be appended after the previous args. So it'd be nice to do so for generality

houseroad replied:
sure, let's make it nice :-)

houseroad committed:
Update on "Support variadic returns in Schema's operator<<"


old: prim::PythonOp(...) ->
new: prim::PythonOp(...) -> ...

Differential Revision: [D16433592](https://our.internmc.facebook.com/intern/diff/D16433592/)

dzhulgakov commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -29,16 +29,22 @@ inline std::ostream& operator<<(std::ostream& out, const FunctionSchema& schema)
-      out << schema.returns()[i].type()->str();
+
+  const auto& returns = schema.returns();
+  out << "(";
```
so the change here to align returning tuple, right? is it why all of the expect files change? are you sure it's a safe change?

I'm still getting confused by how we treat tuple return vs multiple args tbh. cc @suo @zdevito

houseroad replied:
after some digging, I think this diff is fine with existing function schema parser. Here is the reasoning:

1) when parsing the results, we automatically remove the outer pair of parentheses. [related code](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/script/function_schema_parser.cpp#L64) This code also tells us multiple returns should be something like `(int, float)`, one example is `aten::divmod(int x, int y) -> (int, int)` [defined here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/register_prim_ops.cpp#L2723)

2) since we use the outer pair parentheses to represent multi-returns, to return a tuple, we should use two pairs of parentheses. My updates on the expect files are actually correct, because the returns are all single tuples, [one example](https://github.com/pytorch/pytorch/blob/master/test/test_jit.py#L11099)

The test in https://github.com/pytorch/pytorch/pull/23208 already make sure that all the serialized schemas can be correctly parsed (i.e., parsed schemas are exactly equal to the original ones.)

houseroad replied:
also executed the following code:
```
        s1 = parse_schema('any(Tensor self, int dim, bool keepdim=False) -> (Tensor, Tensor)')
        s2 = parse_schema('any(Tensor self, int dim, bool keepdim=False) -> ((Tensor, Tensor))')
        print(len(s1.returns))
        print(len(s2.returns))
```

output is
```
2
1
```

this also shows that, for tuple, we should use '(())'

dzhulgakov replied:
Yeah, existing parsing is a bit non-pythonic (as python would treat it as a tuple). However, I think we can land it avoid changing existing parser and schemas

houseroad committed:
Update on "Support variadic returns in Schema's operator<<"


old: prim::PythonOp(...) ->
new: prim::PythonOp(...) -> ...

Differential Revision: [D16433592](https://our.internmc.facebook.com/intern/diff/D16433592/)
